extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/indexing/MSQCompactionRunner.java

```diff
@@ -0,0 +1,508 @@
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class MSQCompactionRunner implements CompactionRunner
+{
+  private static final Logger log = new Logger(MSQCompactionRunner.class);
+  private static final Granularity DEFAULT_SEGMENT_GRANULARITY = Granularities.ALL;
+
+  private final ObjectMapper jsonMapper;
+  private final Injector injector;
+
+  public static final String TIME_VIRTUAL_COLUMN = "__vTime";
+  public static final String TIME_COLUMN = ColumnHolder.TIME_COLUMN_NAME;
+
+  @JsonIgnore
+  private final CurrentSubTaskHolder currentSubTaskHolder = new CurrentSubTaskHolder(
+      (taskObject, config) -> {
+        final MSQControllerTask msqControllerTask = (MSQControllerTask) taskObject;
+        msqControllerTask.stopGracefully(config);
+      });
+
+
+  public MSQCompactionRunner(@JacksonInject ObjectMapper jsonMapper, @JacksonInject Injector injector)
+  {
+    this.jsonMapper = jsonMapper;
+    this.injector = injector;
+  }
+
+  @Override
+  public NonnullPair<Boolean, String> supportsCompactionSpec(
+      CompactionTask compactionTask
+  )
+  {
+    if (compactionTask.getTuningConfig() != null) {
+      PartitionsSpec partitionsSpec = compactionTask.getTuningConfig().getPartitionsSpec();
+      if (!(partitionsSpec instanceof DynamicPartitionsSpec
+            || partitionsSpec instanceof DimensionRangePartitionsSpec)) {
+        return new NonnullPair<>(false, StringUtils.format(
+            "Invalid partition spec type[%s] for MSQ compaction engine."
+            + " Type must be either DynamicPartitionsSpec or DynamicRangePartitionsSpec.",
+            partitionsSpec.getClass()
+        ));
+      }
+      if (partitionsSpec instanceof DynamicPartitionsSpec
+          && ((DynamicPartitionsSpec) partitionsSpec).getMaxTotalRows() != null) {
+        return new NonnullPair<>(false, StringUtils.format(
+            "maxTotalRows[%d] in DynamicPartitionsSpec not supported for MSQ compaction engine.",
+            ((DynamicPartitionsSpec) partitionsSpec).getMaxTotalRows()
+        ));
+      }
+    }
+    if (compactionTask.getMetricsSpec() != null
+        && compactionTask.getGranularitySpec() != null
+        && !compactionTask.getGranularitySpec()
+                          .isRollup()) {
+      return new NonnullPair<>(
+          false,
+          "rollup in granularitySpec must be set to true if metricsSpec is specifed "
+          + "for MSQ compaction engine."
+      );
+    }
+
+    QueryContext compactionTaskContext = new QueryContext(compactionTask.getContext());
+    if (!MultiStageQueryContext.isFinalizeAggregations(compactionTaskContext)) {
+      return new NonnullPair<>(false, StringUtils.format(
+          "Config[%s] cannot be set to false for auto-compaction with MSQ engine.",
+          MultiStageQueryContext.CTX_FINALIZE_AGGREGATIONS
+      ));
```

Comment:
I fear that it is deviating from the normal behavior of the compaction. Finalize aggregations will finalize the aggregation, and throw away the intermediate data. For example, in the case of sketches, finalizing will save the estimated value in the segment and eliminate the sketch itself. 
Compare this with `rollup = true` in the native engine, where we will also store the sketch data.

This will produce a discrepancy when trying to roll up the data from the segments that have been compacted.

Moreover, w.r.t dimensionToAggregationMap - I think its an unnecessary restriction that can cause trouble. Consider a case where there are two segments with a column - one is created with sketches and `finalizeAggregations = true`, while the other is created where the columns is a dimension. The compaction in native will treat the column as a dimension in both the places, however it will get confusing as to how the column will get treated as in case of MSQ compaction.
We shouldn't store the aggregator used to create a dimension because the intermediate state that the aggregated value stores can get lost when finalizeAggregations = true (depends on the aggregation).

Reply:
I am of the opinion that `finalizeAggregations` shouldn't be set by the users at all. We should figure out the value from the given native compaction spec, and IMO rollup = true means finalizeAggregations = false. We are not allowing rollup = false in MSQ, so we don't have to worry about that.

Reply:
Made changes to have `finalizeAggregations` always set to `false`.